arch/arm/src/stm32wb/stm32wb_spi.c

```diff
@@ -0,0 +1,1819 @@
+
+static inline uint8_t spi_readbyte(struct stm32wb_spidev_s *priv)
+{
+  /* Wait until the receive buffer is not empty */
+
+  while ((spi_getreg(priv, STM32WB_SPI_SR_OFFSET) & SPI_SR_RXNE) == 0);
+
+  /* Then return the received byte */
+
+  return spi_getreg8(priv, STM32WB_SPI_DR_OFFSET);
+}
+
+/****************************************************************************
+ * Name: spi_writeword
+ *
+ * Description:
+ *   Write one 16-bit frame to the SPI FIFO
+ *
+ * Input Parameters:
+ *   priv - Device-specific state data
+ *   byte - Word to send
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static inline void spi_writeword(struct stm32wb_spidev_s *priv,
+                                 uint16_t word)
+{
+  /* Wait until the transmit buffer is empty */
+
+  while ((spi_getreg(priv, STM32WB_SPI_SR_OFFSET) & SPI_SR_TXE) == 0);
+
+  /* Then send the byte */
+
+  spi_putreg(priv, STM32WB_SPI_DR_OFFSET, word);
+}
+
+/****************************************************************************
+ * Name: spi_writebyte
+ *
+ * Description:
+ *   Write one 8-bit frame to the SPI FIFO
+ *
+ * Input Parameters:
+ *   priv - Device-specific state data
+ *   byte - Byte to send
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static inline void spi_writebyte(struct stm32wb_spidev_s *priv,
+                                 uint8_t byte)
+{
+  /* Wait until the transmit buffer is empty */
+
+  while ((spi_getreg(priv, STM32WB_SPI_SR_OFFSET) & SPI_SR_TXE) == 0);
+
+  /* Then send the byte */
+
+  spi_putreg8(priv, STM32WB_SPI_DR_OFFSET, byte);
+}
+
+/****************************************************************************
+ * Name: spi_16bitmode
+ *
+ * Description:
+ *   Check if the SPI is operating in 16-bit mode
+ *
+ * Input Parameters:
+ *   priv     - Device-specific state data
+ *
+ * Returned Value:
+ *   true: 16-bit mode, false: 8-bit mode
+ *
+ ****************************************************************************/
+
+static inline bool spi_16bitmode(struct stm32wb_spidev_s *priv)
+{
+  return (priv->nbits > 8);
+}
+
+/****************************************************************************
+ * Name: spi_dmarxwaitw
+ *
+ * Description:
+ *   Wait for DMA to complete.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_STM32WB_SPI_DMA
+static int spi_dmarxwait(struct stm32wb_spidev_s *priv)
+{
+  int ret;
+
+  /* Take the semaphore (perhaps waiting).  If the result is zero, then the
+   * DMA must not really have completed???
+   */
+
+  do
+    {
+      ret = nxsem_wait_uninterruptible(&priv->rxsem);
+
+      /* The only expected error is ECANCELED which would occur if the
+       * calling thread were canceled.
+       */
+
+      DEBUGASSERT(ret == OK || ret == -ECANCELED);
+    }
+  while (priv->rxresult == 0 && ret == OK);
```

Comment:
As I understand the meaning of this `-ECANCELED` return code, the correct behavior would be to break the waiting loop because the task needs to be cancelled.

Reply:
I reviewed again this part and I think you are right. My main concern was that semaphore could be signalled after waiter exists with `ECANCELED` and the next call will take semaphore with no wait because of the previous call as there is not reset of a semaphore before DMA transaction starts. But logic is a bit more tricky here. The "previous" calls are consumed by `priv->rxresult == 0` condition, so even semaphore wait returns OK we will try to take semaphore again due to other condition, so the only exit criteria from the wait loop is either `ECANCELED` or `priv->rxresult != 0` + `ret == OK`. So the checking of return value makes sense and I we will need to fix part from https://github.com/apache/incubator-nuttx/pull/6492 as well. probably I will do it in the new PR.